### ISSUE TITLE
Inference on CPU

### ISSUE THREAD

**Issue by qinliuliuqin** [facebookresearch/sonata#18]
Dear authors,

Thank you for this great work! I am trying to use your model as an efficient point cloud feature extractor on CPU. However, it fails due to CUDA-only implicit_gemm. Do you have any suggestions or insights on how to run your model on CPU? Thank you very much!

**Comment by Gofinge** [comment]
Looks like the utilization of Sparse Convolution becomes the one restriction of applying Sonata with CPU. Maybe we can train one version with Octree-based Sparse Convolution implemented by ocnn-pytorch. This fully PyTorch-implemented library can run on CPU (Yet a bit slower on GPU).

**Comment by qinliuliuqin** [comment]
@Gofinge Thank you for your prompt response! MinkowskiEngine supports CPU mode for sparse conv, and [3D UNet](https://github.com/xuxw98/ESAM/blob/c5a39477a62ff4f15a0d9bf51d4ba70af71cdac8/oneformer3d/mink_unet.py#L660) implemented by ME works well on CPU. However, it's a little bit large (~500M) and slow on CPU (~400ms CPU latency with 20k points). I was recommended to try Sonata and Point Transformer V3, but it turns out that their current version does not support CPU well. Do you think the CPU version of your model can outperform 3D UNet in terms of model size, computational efficiency, and performance? If so, we may develop some cool 3D applications on edge devices. I’d really appreciate any insights you might have on this.

**Comment by Gofinge** [comment]
Sorry for the late response. Yeah, MinkEngine is too old and hard to install. Yet we do have one option for CPU-ready PTv3: replacing spconv with OCNN-PyTorch. OCNN-PyTorch is fully written with native PyTorch code, yet a bit slower (faster than Mink), but good for CPU inference.

**Comment by vesko233** [comment]
@Gofinge How difficult would it be to replace spconv with OCNN-pytorch in PTv3 ? Does it involve rewriting the entire model so that it supports the Octree object from OCNN-Pytorch ? Just trying to estimate effort here..

**Comment by vesko233** [comment]
Thank you so much @Gofinge ! Glad to hear that you'll be including that in the future

**Comment by dubeamit** [comment]
Hi @Gofinge 
I've tried the above code but I keep getting this error.
```
/opt/conda/conda-bld/pytorch_1728929546833/work/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:144: operator(): block: [115,0,0], thread: [31,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
```
I suppect it's due NAN during loss.

Do you have the hyperparameter values that works with PTv3 using ocnn backbone?

I placed the above code in `Pointcept/pointcept/models/point_transformer_v3/point_transformer_v3m2_ocnn_base.py`
and I'm using the following config.

```
_base_ = ["../_base_/default_runtime.py"]

enable_wandb = False
# misc custom setting
batch_size = 1  # bs: total bs in all gpus
num_worker = 1
mix_prob = 0.8
empty_cache = True
enable_amp = True
clip_grad = 3.0

# model settings
model = dict(
    type="DefaultSegmentorV2",
    num_classes=5,
    backbone_out_channels=64,
    backbone=dict(
        type="PT-v3m2",
        in_channels=6,
        order=("z", "z-trans", "hilbert", "hilbert-trans"),
        stride=(2, 2, 2, 2),
        enc_depths=(2, 2, 2, 6, 2),
        enc_channels=(32, 64, 128, 256, 512),
        enc_num_head=(2, 4, 8, 16, 32),
        enc_patch_size=(1024, 1024, 1024, 1024, 1024),
        dec_depths=(2, 2, 2, 2),
        dec_channels=(64, 64, 128, 256),
        dec_num_head=(4, 4, 8, 16),
        dec_patch_size=(1024, 1024, 1024, 1024),
        mlp_ratio=4,
        qkv_bias=True,
        qk_scale=None,
        attn_drop=0.0,
        proj_drop=0.0,
        drop_path=0.3,
        shuffle_orders=True,
        pre_norm=True,
        enable_rpe=False,
        enable_flash=False,       # since using turing gpu
        upcast_attention=False,
        upcast_softmax=False,
        cls_mode=False,
        pdnorm_bn=False,
        pdnorm_ln=False,
        pdnorm_decouple=True,
        pdnorm_adaptive=False,
        pdnorm_affine=True,
        pdnorm_conditions=("ScanNet", "S3DIS", "Structured3D"),
    ),
    criteria=[
        dict(type="CrossEntropyLoss", loss_weight=1.0, ignore_index=-1),
        dict(type="LovaszLoss", mode="multiclass", loss_weight=1.0, ignore_index=-1),
    ],
)

## If you wanted to train for 100 epochs and evaluate every 10 epochs, you would set:
## epoch = 100
## eval_epoch = 10
## This would set loop = 100 // 10 = 10. The trainer would run for 10 outer epochs, and in each one, it would loop through the data 10 times. Total passes = 10 * 10 = 100.
## scheduler settings
epoch = 100
eval_epoch = 10
optimizer = dict(type="AdamW", lr=0.006, weight_decay=0.05)
scheduler = dict(
    type="OneCycleLR",
    max_lr=[0.006, 0.0006],
    pct_start=0.05,
    anneal_strategy="cos",
    div_factor=10.0,
    final_div_factor=1000.0,
)



param_dicts = [dict(keyword="block", lr=0.0006)]

# dataset settings
dataset_type = "ScanNetDataset"
data_root = "data/nwd_pointcept_data"

data = dict(
    num_classes=5,
    ignore_index=-1,
    names=[
        "wall",
        "floor",
        "door",
        "window",
        "other",
    ],
    train=dict(
        type=dataset_type,
        split="train",
        data_root=data_root,
        transform=[
            dict(type="CenterShift", apply_z=True),
            dict(
                type="RandomDropout", dropout_ratio=0.2, dropout_application_ratio=0.2
            ),
            # dict(type="RandomRotateTargetAngle", angle=(1/2, 1, 3/2), center=[0, 0, 0], axis="z", p=0.75),
            dict(type="RandomRotate", angle=[-1, 1], axis="z", center=[0, 0, 0], p=0.5),
            dict(type="RandomRotate", angle=[-1 / 64, 1 / 64], axis="x", p=0.5),
            dict(type="RandomRotate", angle=[-1 / 64, 1 / 64], axis="y", p=0.5),
            dict(type="RandomScale", scale=[0.9, 1.1]),
            # dict(type="RandomShift", shift=[0.2, 0.2, 0.2]),
            dict(type="RandomFlip", p=0.5),
            dict(type="RandomJitter", sigma=0.005, clip=0.02),
            dict(type="ElasticDistortion", distortion_params=[[0.2, 0.4], [0.8, 1.6]]),
            dict(type="ChromaticAutoContrast", p=0.2, blend_factor=None),
            dict(type="ChromaticTranslation", p=0.95, ratio=0.05),
            dict(type="ChromaticJitter", p=0.95, std=0.05),
            # dict(type="HueSaturationTranslation", hue_max=0.2, saturation_max=0.2),
            # dict(type="RandomColorDrop", p=0.2, color_augment=0.0),
            dict(
                type="GridSample",
                grid_size=0.02,
                hash_type="fnv",
                mode="train",
                return_grid_coord=True,
            ),
            dict(type="SphereCrop", point_max=102400, mode="random"),
            dict(type="CenterShift", apply_z=False),
            dict(type="NormalizeColor"),
            # dict(type="ShufflePoint"),
            dict(type="ToTensor"),
            dict(
                type="Collect",
                keys=("coord", "grid_coord", "segment"),
                feat_keys=("color", "normal"),
            ),
        ],
        test_mode=False,
        # loop=1,
    ),
    val=dict(
        type=dataset_type,
        split="val",
        data_root=data_root,
        transform=[
            dict(type="CenterShift", apply_z=True),
            dict(type="Copy", keys_dict={"segment": "origin_segment"}),
            dict(
                type="GridSample",
                grid_size=0.02,
                hash_type="fnv",
                mode="train",
                return_grid_coord=True,
                return_inverse=True,
            ),
            dict(type="CenterShift", apply_z=False),
            dict(type="NormalizeColor"),
            dict(type="ToTensor"),
            dict(
                type="Collect",
                keys=("coord", "grid_coord", "segment", "origin_segment", "inverse"),
                feat_keys=("color", "normal"),
            ),
        ],
        test_mode=False,
        # loop=1,
    ),
    test=dict(
        type=dataset_type,
        split="val",
        data_root=data_root,
        transform=[
            dict(type="CenterShift", apply_z=True),
            dict(type="NormalizeColor"),
        ],
        test_mode=True,
        test_cfg=dict(
            voxelize=dict(
                type="GridSample",
                # grid_size=0.1,
                grid_size=0.02,
                hash_type="fnv",
                mode="test",
                return_grid_coord=True,
            ),
            crop=None,  # Default
            post_transform=[
                dict(type="CenterShift", apply_z=False),
                dict(type="ToTensor"),
                dict(
                    type="Collect",
                    keys=("coord", "grid_coord", "index"),
                    feat_keys=("color", "normal"),
                ),
            ],
            aug_transform=[ 
                [dict(type="RandomRotateTargetAngle", angle=[0], axis="z", center=[0, 0, 0], p=1)],
            ],
        ),
    ),
)



```
